rpc/notification.go

```diff
@@ -0,0 +1,299 @@
+func (n *bufferedNotifier) send(id string, data interface{}) error {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	if n.stopped {
+		return notifierStoppedErr
+	}
+
+	var (
+		subscription *bufferedSubscription
+		found        bool
+	)
+
+	// check if subscription is associated with this connection, it might be cancelled
+	// (subscribe/connection closed)
+	if subscription, found = n.subscriptions[id]; !found {
+		glog.V(logger.Error).Infof("received notification for unknown subscription %s\n", id)
+		return NotificationNotFoundErr
+	}
+
+	// don't accept new notifications for this subscription
+	if data == unsubSignal {
+		delete(n.subscriptions, id)
+		if subscription.unsub != nil {
+			subscription.once.Do(func() { subscription.unsub(id) })
+		}
+	}
+
+	subscription.lastNotification = time.Now()
+
+	if len(n.queue) >= n.queueSize {
+		glog.V(logger.Warn).Infoln("too many buffered notifications -> close connection")
+		n.codec.Close()
+		return notificationQueueFullErr
+	}
+
+	n.queue <- &notification{subscription, data}
+	return nil
+}
+
+// run reads notifications from the internal queue and sends them to the client. In case of an
+// error, or when the codec is closed it will cancel all active subscriptions and returns.
+func (n *bufferedNotifier) run() {
+	defer func() {
+		n.mu.Lock()
+		defer n.mu.Unlock()
+
+		n.stopped = true
+		close(n.queue)
+
+		// on exit call unsubscribe callback
+		for id, sub := range n.subscriptions {
+			if sub.unsub != nil {
+				sub.once.Do(func() { sub.unsub(id) })
+			}
+			close(sub.flushed)
+			delete(n.subscriptions, id)
+		}
+	}()
+
+	// check and remove inactive subscriptions every minute
+	cleanupTimer := time.NewTicker(1 * time.Minute)
+
+	for {
+		select {
+		case notification := <-n.queue:
+
+			// It can happen that an event is raised before the RPC server was able to send the sub
+			// id to the client. Therefore subscriptions are marked as pending until the sub id was
+			// send. The RPC server will activate the subscription by closing the pending chan.
+			<-notification.sub.pending
+
+			if notification.data == unsubSignal {
+				// unsubSignal is the last accepted message for this subscription. Raise the signal
+				// that all buffered notifications are send by closing the flushed channel.
+				close(notification.sub.flushed)
+			} else {
+				msg := n.codec.CreateNotification(notification.sub.id, notification.data)
+				if err := n.codec.Write(msg); err != nil {
+					n.codec.Close()
+					// unable to send notification to client, unsubscribe all subscriptions
+					glog.V(logger.Warn).Infof("unable to send notification - %v\n", err)
+					return
+				}
+			}
+		case <-n.codec.Closed(): // connection was closed
+			glog.V(logger.Debug).Infoln("codec closed, stop subscriptions")
+			return
+		case <-cleanupTimer.C:
+			n.cleanup()
+		}
+	}
+}
+
+// cleanup removes subscriptions that have not generated a notification in the last 5 minutes.
+func (n *bufferedNotifier) cleanup() {
```

Comment:
I'm not sure about this method or the concept behind it to be honest. If I understand correctly, you drop a subscription if it idles for more than 5 minutes. But that wouldn't make much sense since many subscriptions will idle almost all the time and only occasionally get pinged (e.g. inbound transaction to my account). Furthermore if the client is not notified of the drop, then it won't even realize that the subscription's long dead.

Could you elaborate a bit on why this was needed and what scenario it should be solving?